### PR TITLE
drivers: modem: gsm_ppp: proper attach retry

### PR DESCRIPTION
**Pull request by endian-benjamin**
In a previous commit, a check was added to ensure modem was properly
attached to packet service before initializing PPP. Failure to perform
this check can lead to the dreaded 'NO CARRIER' issue.

While this is a nice idea, the implementation was lacking because when
the check failed, the entire modem initialization procedure was
restarted from square one. For modems/bearers that were slow to attach,
this is potentially disastrous. The proper solution is to loop only the
'AT+CGATT?' part until it succeeds, or fails N times.

This commit implements this looping behavior (using the work queue).

Signed-off-by: Benjamin Lindqvist <benjamin.lindqvist@endian.se>

### CODE REVIEW

File: drivers/modem/gsm_ppp.c
Comment (jukkar) [inline drivers/modem/gsm_ppp.c:352]: Could you add this retry counter to `gsm_modem` context, as then we could (at least in theory) have multiple modem devices in the system.

File: drivers/modem/gsm_ppp.c
Comment (jukkar) [inline drivers/modem/gsm_ppp.c:404]: Having this configurable would be great, it should not be too big change to add that to this PR too

**Review by jukkar — COMMENTED** [review]
Looks reasonable, couple minor comments

File: drivers/modem/gsm_ppp.c
Comment (mniestroj) [inline drivers/modem/gsm_ppp.c:414]: Why do you increase that?

File: drivers/modem/gsm_ppp.c
Comment (mniestroj) [inline drivers/modem/gsm_ppp.c:402]: Where does 60s come from? I don't see it explicitly in the code, which means that this comment might quickly get outdated.

File: drivers/modem/gsm_ppp.c
Comment (endian-benjamin) [reply drivers/modem/gsm_ppp.c]: Not sure, this patch has been on the shelf for a little while now. I believe it makes more sense with the latest force push.

File: drivers/modem/gsm_ppp.c
Comment (endian-benjamin) [reply drivers/modem/gsm_ppp.c]: Agree, see latest force push

File: drivers/modem/Kconfig.gsm
Comment (jukkar) [inline drivers/modem/Kconfig.gsm:57]: Hmm, the prompt talks about retries, but the option name and the help text talks about time (seconds). This looks confusing now, perhaps the prompt should talk about time too.

**Comment by endian-benjamin** [comment]
@jukkar Feel free to comment on code but don't smash the merge button before I test the newest force push

File: drivers/modem/gsm_ppp.c
Comment (jukkar) [inline drivers/modem/gsm_ppp.c:69]: Please put this before the bitfields so we might avoid a hole in the memor

**Review by jukkar — COMMENTED** [review]
Looks good, couple minor adjustments still needed

**Comment by jukkar** [comment]
> Feel free to comment on code but don't smash the merge button before I test the newest force push

Sure, note also that you should be able to add DNM label if you wish that the PR is not yet merged, and then remove the label when you are ok with it.

**Comment by endian-benjamin** [comment]
@jukkar this works as expected on a simcom 7600e lte modem.

**Review by jukkar — APPROVED** [review]

File: drivers/modem/gsm_ppp.c
Comment (mniestroj) [inline drivers/modem/gsm_ppp.c:404]:
```
...attach_retries = ...ATTACH_TIMEOUT;
```
This seems wrong. Either there is a timeout in seconds or there is a retry limit.

File: drivers/modem/gsm_ppp.c
Comment (mniestroj) [inline drivers/modem/gsm_ppp.c:396]: This is a whitespace change only, so please exclude it from this commit.

**Review by mniestroj — CHANGES_REQUESTED** [review]

File: drivers/modem/gsm_ppp.c
Comment (mniestroj) [reply drivers/modem/gsm_ppp.c:406]: >Well it retries at 1 Hz so they coincide.

If someone will change `K_SECONDS(1)` to some other value, then above statement will be wrong. Maybe use something like:
```
#define GSM_ATTACH_RETRY_DELAY_MSEC 1000
```
then in the code
```
gsm->attach_retries = CONFIG_MODEM_GSM_ATTACH_TIMEOUT * MSEC_PER_SEC / GSM_ATTACH_RETRY_DELAY_MSEC;
```
and then
```
void)k_delayed_work_submit(&gsm->gsm_configure_work,
			K_MSEC(GSM_ATTACH_RETRY_DELAY_MSEC));
```

Comment (endian-benjamin) [reply drivers/modem/gsm_ppp.c]: yeah looks clean enough I suppose, how about now? I'll squash as soon as it's approved

Comment (mniestroj) [reply drivers/modem/gsm_ppp.c]: I like it now. Please squash and I will approve this PR later on, so that we don't merge accidentally. Thanks!

Comment (endian-benjamin) [reply drivers/modem/gsm_ppp.c]: ping

**Review by mniestroj — APPROVED** [review]

**Pull request merged**
